service/controller/app/v1/kubeconfig/kubeconfig.go

```diff
@@ -0,0 +1,107 @@
+package kubeconfig
+
+import (
+	"context"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/application/v1alpha1"
+	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
+	kubeconfiglib "github.com/giantswarm/kubeconfig"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/giantswarm/app-operator/service/controller/app/v1/key"
+)
+
+// Config represents the configuration used to create a new kubeconfig service.
+type Config struct {
+	G8sClient versioned.Interface
```

Comment:
I can assume that these `G8sClient` and `K8sClient` should the same as what `chart` resource contains itself, doesn't it? 
https://github.com/giantswarm/app-operator/blob/a9ce6eb4c31200c73be1cdec814559fec0e403ec/service/controller/app/v1/resource/chart/resource.go#L37-L38
Rather than copy clients to another instance, what about returning nil but error when Secret is not placed in app CR?

Reply:
The problem is this logic will be needed in 3 resources (chart, configmap and secret). Rather than repeat it in all 3 resources the kubeconfig service handles making sure the correct clientset is used.